apps/aecore/src/aec_vm_chain.erl

```diff
@@ -0,0 +1,140 @@
+        Err = {error, _} -> Err
+    end.
+
+%% @doc Call another contract.
+-spec call_contract(pubkey(), non_neg_integer(), non_neg_integer(), binary(),
+                    [non_neg_integer()], chain_state()) ->
+        {ok, aec_vm_chain_api:call_result(), chain_state()} | {error, term()}.
+call_contract(Target, Gas, Value, CallData, CallStack,
+              State = #state{ trees   = Trees,
+                              height  = Height,
+                              account = ContractKey,
+                              nonce   = Nonce }) ->
+    VmVersion = 0,  %% TODO
+    {ok, CallTx} =
+        aect_call_tx:new(#{ caller     => ContractKey,
+                            nonce      => Nonce,
+                            contract   => Target,
+                            vm_version => VmVersion,
+                            fee        => 0,
+                            amount     => Value,
+                            gas        => Gas,
+                            gas_price  => 0,
+                            call_data  => CallData,
+                            call_stack => CallStack }),
+    case aetx:check_from_contract(CallTx, Trees, Height) of
+        Err = {error, _} -> Err;
+        {ok, Trees1} ->
+            {ok, Trees2} = aetx:process_from_contract(CallTx, Trees1, Height),
+            CallId  = aect_call:id(ContractKey, Nonce, Target),
+            Call    = aect_state_tree:get_call(Target, CallId, aec_trees:contracts(Trees2)),
+            GasUsed = aect_call:gas_used(Call),
+            Result  = case aect_call:return_value(Call) of
+                          %% TODO: currently we don't set any sensible return value on exceptions
+                          <<>> -> aec_vm_chain_api:call_exception(out_of_gas, GasUsed);
+                          Bin when is_binary(Bin) ->
+                            aec_vm_chain_api:call_result(Bin, GasUsed)
+                      end,
+            {ok, Result, State#state{ trees = Trees2, nonce = Nonce + 1 }}
+    end.
+
+
+%% -- Internal functions -----------------------------------------------------
+
+do_get_balance(ContractKey, Trees) ->
+    ContractsTree = aec_trees:contracts(Trees),
+    Contract      = aect_state_tree:get_contract(ContractKey, ContractsTree),
+    aect_contracts:balance(Contract).
+
+%% TODO: can only spend to proper accounts. Not other contracts.
+%% Note that we cannot use an aec_spend_tx here, since we are spending from a
+%% contract account and not a proper account.
+do_spend(Recipient, ContractKey, Amount, Trees, Height) ->
+    try
+        ContractsTree = aec_trees:contracts(Trees),
+        AccountsTree  = aec_trees:accounts(Trees),
+        Contract      = aect_state_tree:get_contract(ContractKey, ContractsTree),
+        Balance       = aect_contracts:balance(Contract),
+        [ throw(no_funds) || Balance < Amount ],
+        RecipAccount  =
+            case aec_accounts_trees:lookup(Recipient, AccountsTree) of
```

Comment:
This is not quite right, I think. The spend transaction actually creates non-existing accounts using  `aec_trees:ensure_account_at_height/3`. The call is very badly placed (in `aec_spend_tx:check/4`) so it is easy to miss.

Reply:
I pushed a fix on this.